subprojects/core-api/src/main/java/org/gradle/api/file/FileSystemLayout.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.file;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.provider.Provider;
+
+import java.io.File;
+
+/**
+ * Provides provider-based access to file system locations.
+ *
+ * @since 8.5
+ */
+@Incubating
+public interface FileSystemLayout {
```

Comment:
❓ I want to strongly question the need for the new interface.
My main question is what purpose does it serve to the user?

As it is now, it expresses something partial. It claims to be a layout, but it does not provide access to the actual entry-point of the layout. Currently, it allows creating (relative) files "somewhere".

Project layout, though not explicitly mentioned in the javadocs, creates files relative to the project directory. It additionally contains a mutable build directory property, because it is so often accessed. This makes it suitable for the layout definition.

Settings layout, by itself, serves the same purpose -- creating files relative to the settings directory. (We should definitely make it explicit in the documentation.) It also provides the path to the root directory, because there is 1-to-1 correspondence. These things make it a layout as well.

The `FileSystemLayout`, currently, is effectively a factory with a hidden state. Because of this, it is hard for me to imagine that instances of it will be used as a "layout" in isolation from the actual instances of the `ProjectLayout` and `SettingsLayout`. Hence, I don't see a need for an extra abstraction in the API.

I see two approaches:
* (a) Get rid of the `FileSystemLayout` interface, and keep the `file`, `dir`, etc methods on the project and settings-specific interfaces, and make their javadocs explicit about the relative path case (which is definitely the primary use-case). The implementation can still make use of an `AbstractFileSystemLayout`, but this is not visible to the users.
* (b) Keep the `FileSystemLayout`, but then add a method that returns the entry-point of the layout, so it is identifiable, and has a higher chance of getting reused in the plugin/user code. The `ProjectLayout` and `SettingsLayout` should then override this method, explicitly showing which directory is the entry-point.

I am in favor of (a), because it does not introduce new public abstractions for their own sake. Moreover, we still have an option to do (b) at any later point if there is a need. It would also be more natural to document the `file` and `dir` methods on the corresponding layouts to be more specific about the relative file case.

Reply:
I went with option b, Alex. Motivations:

1. SettingsLayout and ProjectLayout are meant mostly for the same need (locations relative to some reference location), mostly dfifer in having  a different call context.
2. They both resolve locations based on some reference/anchor location
3. FileSystemnLayout now introduces the idea of a reference location. I felt tempted to add a method that returned said location, but opted for a softer approach, only describing it as an underlying concept.

Reply:
That's alright.